src/common/sec_websocket_extensions.rs

```diff
@@ -0,0 +1,369 @@
+use std::convert::TryFrom;
+
+use bytes::BytesMut;
+use http::header::SEC_WEBSOCKET_EXTENSIONS;
+
+use util::{Comma, FlatCsv, HeaderValueString, SemiColon};
+use {Error, Header, HeaderValue};
+
+/// `Sec-WebSocket-Extensions` header, defined in [RFC6455][RFC6455_11.3.2]
+///
+/// The `Sec-WebSocket-Extensions` header field is used in the WebSocket
+/// opening handshake.  It is initially sent from the client to the
+/// server, and then subsequently sent from the server to the client, to
+/// agree on a set of protocol-level extensions to use for the duration
+/// of the connection.
+///
+/// ## ABNF
+///
+/// ```text
+/// Sec-WebSocket-Extensions = extension-list
+/// extension-list = 1#extension
+/// extension = extension-token *( ";" extension-param )
+/// extension-token = registered-token
+/// registered-token = token
+/// extension-param = token [ "=" (token | quoted-string) ]
+/// ```
+///
+/// ## Example Values
+///
+/// * `permessage-deflate` (defined in [RFC7692][RFC7692_7])
+/// * `permessage-deflate; server_max_window_bits=10`
+/// * `permessage-deflate; server_max_window_bits=10, permessage-deflate`
+///
+/// ## Example
+///
+/// ```rust
+/// # extern crate headers;
+/// use headers::SecWebsocketExtensions;
+///
+/// let extensions = SecWebsocketExtensions::from_static("permessage-deflate");
+/// ```
+///
+/// ## Splitting and Combining
+///
+/// Note that `Sec-WebSocket-Extensions` may be split or combined across multiple headers.
+/// The following are equivalent:
+/// ```text
+/// Sec-WebSocket-Extensions: foo
+/// Sec-WebSocket-Extensions: bar; baz=2
+/// ```
+/// ```text
+/// Sec-WebSocket-Extensions: foo, bar; baz=2
+/// ```
+///
+/// `SecWebsocketExtensions` splits extensions when decoding and combines them into a single
+/// value when encoding.
+///
+/// [RFC6455_11.3.2]: https://tools.ietf.org/html/rfc6455#section-11.3.2
+/// [RFC7692_7]: https://tools.ietf.org/html/rfc7692#section-7
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SecWebsocketExtensions(pub Vec<WebsocketExtension>);
```

Comment:
We should keep the fields private, so we can change the internal implementation whenever desired.

Reply:
I explained the reason behind this in https://github.com/hyperium/headers/pull/88#discussion_r711390478